Remove Lodash from sites selectors

jsnajdr opened the pull request:
Removes Lodash from `state/sites/selectors` and some sites-related `state/selectors`.

**How to test:**
Verify that unit tests pass.

jsnajdr committed:
Remove Lodash from sites selectors

tyxla commented on client/state/selectors/get-network-sites.js:
```diff
@@ -26,10 +20,9 @@ export default createSelector(
 		}
 
-		return filter(
-			getSitesItems( state ),
```
On par with #53328, we should likely move `getSitesItems` and `isMainSiteOf` to the site selectors directory. Then we can freely move this one there, too (`getNetworkSites`).

jsnajdr replied:
Yes, `getSitesItems` and `getRawSite` are actually already prepared in my pipeline 🙂 But doing it all at once leads to monster PRs.

tyxla replied:
Wow, nice. Yeah, definitely makes sense to do in another PR.

tyxla commented on client/state/selectors/is-eligible-for-dotcom-checklist.js:
```diff
@@ -33,7 +28,7 @@ export default function isEligibleForDotcomChecklist( state, siteId ) {
 
 	// Checklist should not show up if the site is created before the feature was launched.
-	if ( get( siteOptions, 'created_at', '' ) < '2018-02-01' ) {
+	if ( siteOptions?.created_at ?? '' < '2018-02-01' ) {
```
Unrelated to the PR: `'' < '2018-02-01'` will always be `true`. I hope this is intentional.

jsnajdr replied:
It's a behavior that already existed before this PR. And it looks correct: if the `siteId` is not present or doesn't have the `options.created_at` property, the selector will return `false`. And our universal convention is to return `false` or `null` if the answer to the selector's question is "I don't know".

jsnajdr replied:
Wow, there was an insidious bug here caught by unit tests: the `<` operator has higher precedence than `??`, so it was evaluated as `siteOptions?.createdAt ?? ( '' < '2018-02-01' )`. Fixing by adding parens.

tyxla commented on client/state/selectors/is-mapped-domain-site.js:
```diff
@@ -26,8 +20,5 @@ export default function isMappedDomainSite( state, siteId ) {
-		get( site, 'options.is_mapped_domain', false ) &&
-		some( domains, ( { isWPCOMDomain } ) => ! isWPCOMDomain )
-	);
+	return site.options?.is_mapped_domain && domains.some( ( { isWPCOMDomain } ) => ! isWPCOMDomain );
```
Nice shortening here. We don't need a fallback `false` in a boolean condition 👍

jsnajdr replied:
I had to fix the unit test here. It has a wrong mock for `sites.domains`. Instead of returning an array, it returned a `{ isWPCOMDomain: true }` object. And `_.some` then iterated over the `true` value. Using `Array.prototype.some` revealed that.

I changed the mock to an array, and the `isWPCOMDomain` is `false` now. Previously, `true.isWPCOMDomain` evaluated to a falsy `undefined` value.

tyxla commented on client/state/selectors/is-site-failed-migration-source.js:
```diff
@@ -17,7 +12,5 @@ import getRawSite from 'calypso/state/selectors/get-raw-site';
-	const siteMigrationMeta = get( site, 'site_migration', {} );
-
-	return !! get( siteMigrationMeta, 'failed_backup_source', false );
+	return site?.site_migration?.failed_backup_source ?? false;
```
Is it guaranteed that `failed_backup_source` will either be boolean or nullish? If yes, then it's fine to remove the boolean cast.

jsnajdr replied:
Good question and the answer is yes. I verified in the backend PHP source that `failed_backup_source` is either `true`, or not present at all.

tyxla commented on client/state/selectors/user-has-any-atomic-sites.js:
```diff
@@ -17,6 +12,7 @@ import isAtomicSite from 'calypso/state/selectors/is-site-automated-transfer';
 export default createSelector(
-	( state ) => some( getSitesItems( state ), ( site ) => isAtomicSite( state, site.ID ) ),
+	( state ) =>
+		Object.values( getSitesItems( state ) ).some( ( site ) => isAtomicSite( state, site.ID ) ),
```
FWIW here we only need the site IDs, so we could actually go with `Object.keys()` and use the IDs directly.

jsnajdr replied:
Changing this to `Object.keys( ... )`. The difference is that in the `Object.keys( ... )` array, the IDs are strings. While `site.ID` is a number. But apparently it doesn't matter, because there are places that use `Object.keys( ... )` already.

tyxla commented on client/state/sites/selectors/build-seo-title.js:
```diff
@@ -1,17 +1,9 @@
 export default ( titleFormats, type, { site, post = {}, tag = '', date = '' } ) => {
 	const processPiece = ( piece = {}, data ) =>
-		'string' === piece.type ? piece.value : get( data, piece.type, '' );
+		'string' === piece.type ? piece.value : data[ piece.type ] ?? '';
```
Any chance for `data` to be `undefined`?

jsnajdr replied:
`data` is always defined. The `processPiece` and `buildTitle` functions are local and it's easy to verify that all calls include a valid `data` object.

tyxla commented on client/state/sites/selectors/build-seo-title.js:
```diff
@@ -1,17 +1,9 @@
-			( title, piece ) => title + processPiece( piece, data ),
-			''
-		);
+		( titleFormats[ format ] ?? [] ).map( ( piece ) => processPiece( piece, data ) ).join( '' );
```
Any chance for `titleFormats` to be `undefined`?

jsnajdr replied:
By tracing the calls (there are two or three in total) I see that the worst case is that `titleFormats` is an empty `{}` object. Never `null` or `undefined`.

tyxla commented on client/state/sites/selectors/build-seo-title.js:
```diff
@@ -27,15 +19,17 @@ export default ( titleFormats, type, { site, post = {}, tag = '', date = '' } )
-				} ) || get( post, 'title', '' )
+					postTitle: post.title ?? '',
+				} ) ||
+				post.title ||
```
It's odd that sometimes we use nullish coalescing and sometimes we use logical OR for the same thing 🤔

jsnajdr replied:
We can't use `??` everywhere because `buildTitle` returns empty `''` string when it 'fails'. We could replace the `post.title ?? ''` you're commenting on with `post.title || ''`. But it doesn't have a practical impact, it's a matter of aesthetics.

tyxla commented on client/state/sites/selectors/can-jetpack-site-auto-update-files.js:
```diff
@@ -30,7 +25,7 @@ export default function canJetpackSiteAutoUpdateFiles( state, siteId ) {
 	const fileModDisabled = getSiteOption( state, siteId, 'file_mod_disabled' );
 
-	if ( fileModDisabled && includes( fileModDisabled, 'automatic_updater_disabled' ) ) {
+	if ( fileModDisabled && fileModDisabled.includes( 'automatic_updater_disabled' ) ) {
```
Maybe we can just optional chaining and remove the `fileModDisabled &&`?

jsnajdr replied:
I tried it and it crashed. Because `fileModDisabled` can be `false`.

tyxla commented on client/state/sites/selectors/get-jetpack-site-update-files-disabled-reasons.js:
```diff
@@ -31,8 +30,8 @@ export default function getJetpackSiteUpdateFilesDisabledReasons(
 
-	return compact(
-		fileModDisabled.map( ( clue ) => {
+	return fileModDisabled
```
It's weird that we don't check if `fileModDisabled` exists before calling `.map()` on it. We seem to be checking it in other instances in this PR. Should we do it here as well?

jsnajdr replied:
That's true, and the reason why it never crashed for anyone is that the selector was never used since it was added 5 years ago. Instead, we're using the similar `getSiteFileModDisableReason( site )` function. I'll remove the selector.

tyxla commented on client/state/sites/selectors/get-seo-title-formats-for-site.js:
```diff
@@ -17,7 +12,4 @@ import { fromApi as seoTitleFromApi } from 'calypso/components/seo/meta-title-ed
-	seoTitleFromApi,
-	partialRight( get, 'options.advanced_seo_title_formats', {} )
-);
+export default ( site ) => seoTitleFromApi( site?.options?.advanced_seo_title_formats ?? {} );
```
😞  This isn't really a selector. Seems like it would be a good idea to move it out of this directory.

Also, it feels off to have to pass the entire site object when we can just pass the ID and get the rest out of the state. Maybe this is worth refactoring to an actual selector?

jsnajdr replied:
That's true, and `buildSeoTitle` is also not a selector. I'd rather keep this SEO code as is and maybe refactor it on some other occasion.

tyxla replied:
This PR is pretty large already, so sounds fine by me.

tyxla commented on client/state/sites/selectors/get-seo-title.js:
```diff
@@ -1,16 +1,11 @@
 
 export default ( state, type, data ) => {
-	if ( ! has( data, 'site.ID' ) ) {
+	if ( data?.site?.ID ) {
```
Did you mean to preserve the `!`?

jsnajdr replied:
Of course, thanks for spotting this 🙂

tyxla commented on client/state/sites/selectors/get-site-by-slug.js:
```diff
@@ -19,6 +14,6 @@ import getSiteSlug from './get-site-slug';
 export default createSelector(
 	( state, siteSlug ) =>
-		find( getSitesItems( state ), ( site ) => getSiteSlug( state, site.ID ) === siteSlug ) || null,
+		getSitesItems( state ).find( ( site ) => getSiteSlug( state, site.ID ) === siteSlug ) ?? null,
```
Since `getSitesItems()` will return an object, should we wrap it in a `Object.values()` before calling `.find()` on it?

And since we actually only need the ID of the site, we could use `Object.keys()`.

jsnajdr replied:
Good catch, this actually crashed Calypso at runtime. Fixing.

jsnajdr replied:
> And since we actually only need the ID of the site, we could use `Object.keys()`.

In this case we need `Object.values()` because the result of `.find` is supposed to be the site object.

tyxla commented on client/state/sites/selectors/get-site-collisions.js:
```diff
@@ -14,21 +12,18 @@ import getSitesItems from 'calypso/state/selectors/get-sites-items';
+			const siteUrlSansProtocol = withoutHttp( site.URL );
+			return (
+				! site.jetpack &&
+				sitesItems.some(
+					( jetpackSite ) =>
+						jetpackSite.jetpack && siteUrlSansProtocol === withoutHttp( jetpackSite.URL )
+				)
```
Do we really want to execute this for all `sitesItems`? I think we can move this to a separate constant above and reuse it here.

jsnajdr replied:
What exactly could be moved to a constant? I don't see any opportunity to do that. `siteUrlSansProtocol` is different for each iteration.

tyxla replied:
Ha, you're right, that was an oversight. Disregard.

tyxla commented on client/state/sites/selectors/get-site-plan-slug.js:
```diff
@@ -1,13 +1,8 @@
 
 export default function getSitePlanSlug( state, siteId ) {
-	return get( getSitePlan( state, siteId ), 'product_slug' );
+	return getSitePlan( state, siteId )?.product_slug ?? null;
```
Are we intentionally changing the default value from `undefined` to `null`?

jsnajdr replied:
Yes, this was intentional -- it's common for Calypso selectors to return `null` if the result means "I don't know".

tyxla replied:
👍  I wanted to make sure we've weighted any potential impact this has on the existing code that uses this selector.

jsnajdr replied:
Now I see there are actually two `getSitePlanSlug` selector. The other one in `sites/plans/selectors` is slightly different, and returns `null` in the "not found" case 🙈

tyxla reviewed (commented):
It's wonderful to see so much Lodash removal ❤️ 

It's a rather big PR, I've left some questions and thoughts, let me know what you think.

jsnajdr committed:
Address review feedback

jsnajdr committed:
Remove the unused getJetpackSiteUpdateFilesDisabledReasons selector

jsnajdr commented:
@tyxla I addressed all review feedback and did a lot of fixes. Also removed the `getJetpackSiteUpdateFilesDisabledReasons` selector.

Please have another look. And thanks for reviewing thoroughly. Turns out that the v1 of this PR had a lot of bugs.

tyxla commented on client/state/selectors/is-eligible-for-dotcom-checklist.js:
```diff
@@ -33,7 +28,7 @@ export default function isEligibleForDotcomChecklist( state, siteId ) {
 
 	// Checklist should not show up if the site is created before the feature was launched.
-	if ( get( siteOptions, 'created_at', '' ) < '2018-02-01' ) {
+	if ( ( siteOptions?.created_at ?? '' ) < '2018-02-01' ) {
```
🥇

tyxla commented on client/state/selectors/test/is-mapped-domain-site.js:
```diff
@@ -22,9 +22,11 @@ describe( '#isMappedDomainSite()', () => {
-				[ siteId ]: {
-					isWPCOMDomain: true,
-				},
+				[ siteId ]: [
```
Nice catch 👍

tyxla reviewed (approved):
The code looks great 👍  Also smoke testing didn't reveal any problems. Great work 👏
